File: src/services/errors/appError.ts
/**
 * Application error taxonomy (Phase 3).
 *
 * One error shape for ALL service layers. Rules:
 * - `category` identifies the boundary that failed (never raw subsystem text),
 * - `code` is a stable machine-readable string,
 * - `recoverable` tells the caller whether retry/recovery is appropriate,
 * - `userMessage` is the ONLY string UI/log surfaces may show to users —
 *   it must never contain SQL, paths, credentials, or domain values,
 * - `cause` carries internal diagnostics for logging only (logger redacts).
 *
 * PersistenceError (Phase 2) keeps its own shape for the data layer; services
 * translate it into AppError at the boundary (see `fromPersistence`).
 */

import { PersistenceError } from '../../data/database/errors.ts';

export type AppErrorCategory =
  | 'service'
  | 'persistence'
  | 'filesystem'
  | 'notification'
  | 'permission'
  | 'security'
  | 'validation'
  | 'media'
  | 'unknown';

export interface AppErrorOptions {
  recoverable?: boolean;
  userMessage?: string;
  cause?: unknown;
}

export class AppError extends Error {
  readonly category: AppErrorCategory;
  readonly code: string;
  readonly recoverable: boolean;
  readonly userMessage: string;
  readonly cause?: unknown;

  constructor(category: AppErrorCategory, code: string, options: AppErrorOptions = {}) {
    super(`${category}:${code}`);
    this.name = 'AppError';
    this.category = category;
    this.code = code;
    this.recoverable = options.recoverable ?? true;
    this.userMessage = options.userMessage ?? DEFAULT_USER_MESSAGES[category];
    this.cause = options.cause;
  }
}

const DEFAULT_USER_MESSAGES: Record<AppErrorCategory, string> = {
  service: 'A service failed.',
  persistence: 'Local data could not be accessed.',
  filesystem: 'A file operation failed.',
  notification: 'Notification scheduling failed.',
  permission: 'A required permission is missing.',
  security: 'A security operation failed.',
  validation: 'Input is invalid.',
  media: 'Media processing failed.',
  unknown: 'An unexpected error occurred.',
};

/** Stable user-facing message — never raw internals. */
export function safeUserMessage(error: unknown): string {
  if (error instanceof AppError) return error.userMessage;
  if (error instanceof PersistenceError) return DEFAULT_USER_MESSAGES.persistence;
  return DEFAULT_USER_MESSAGES.unknown;
}

/** Wrap any failure as AppError; pass AppError through untouched. */
export function normalizeAppError(
  cause: unknown,
  category: AppErrorCategory,
  code: string,
  options: AppErrorOptions = {},
): AppError {
  if (cause instanceof AppError) return cause;
  if (cause instanceof PersistenceError) return fromPersistence(cause, code);
  return new AppError(category, code, { ...options, cause });
}

/** Translates a Phase 2 PersistenceError without leaking internals. */
export function fromPersistence(error: PersistenceError, code = 'persistence-failed'): AppError {
  const recoverable = error.code !== 'init-failed' && error.code !== 'migration-failed';
  return new AppError('persistence', code, { recoverable, cause: error });
}

File: app/src/main/java/com/twohearts/app/services/error/AppError.kt
package com.twohearts.app.services.error

/**
 * AppError — unified error type matching legacy AppError taxonomy.
 *
 * Every error in the app flows through this class, providing:
 * - Category (validation, persistence, permission, etc.)
 * - Code (machine-readable identifier)
 * - Safe message (user-facing, no internal details)
 * - Original exception (for logging)
 *
 * Matches legacy error taxonomy exactly.
 */
data class AppError(
    val category: ErrorCategory,
    val code: String,
    val safeMessage: String,
    val originalException: Throwable? = null
) : Exception(safeMessage, originalException) {

    companion object {
        // Validation errors
        fun validation(code: String, message: String) = AppError(
            category = ErrorCategory.VALIDATION,
            code = code,
            safeMessage = message
        )

        // Persistence errors
        fun persistence(code: String, message: String, cause: Throwable? = null) = AppError(
            category = ErrorCategory.PERSISTENCE,
            code = code,
            safeMessage = message,
            originalException = cause
        )

        // Permission errors
        fun permission(code: String, message: String) = AppError(
            category = ErrorCategory.PERMISSION,
            code = code,
            safeMessage = message
        )

        // Not found errors
        fun notFound(code: String, entity: String) = AppError(
            category = ErrorCategory.NOT_FOUND,
            code = code,
            safeMessage = "$entity not found"
        )

        // Security errors
        fun security(code: String, message: String) = AppError(
            category = ErrorCategory.SECURITY,
            code = code,
            safeMessage = message
        )

        // Media errors
        fun media(code: String, message: String, cause: Throwable? = null) = AppError(
            category = ErrorCategory.MEDIA,
            code = code,
            safeMessage = message,
            originalException = cause
        )

        // Unknown errors
        fun unknown(message: String, cause: Throwable? = null) = AppError(
            category = ErrorCategory.UNKNOWN,
            code = "UNKNOWN",
            safeMessage = message,
            originalException = cause
        )
    }
}

/**
 * Error categories matching legacy AppError categories.
 */
enum class ErrorCategory {
    VALIDATION,
    PERSISTENCE,
    PERMISSION,
    NOT_FOUND,
    SECURITY,
    MEDIA,
    NETWORK,
    UNKNOWN
}
